### PR TITLE
fix #14 

### LINKED ISSUE #14: The provided typescript typings are suboptimal and do not work
PR #15 

The typings introduced with #10 are suboptimal and do not work in that they do now allow typescript to infer the prototype/inheritance chain properly during static code analysis.

One result of that is that the mandatory name and message fields are no longer declared, not to mention the fact that AssertionErrorConstructor#new() must return either a class or interface but not a type.

E.g. given the following code

```
import AssertionError = require('assertion-error');

class MySpecialAssertionError extends AssertionError {

  public constructor(message: string) {
    super(message);
  }
}

const errors: Error[] = [];

errors.push(new MySpecialAssertionError('very special'));
```

will fail because typescript will complain about MySpecialAssertionError not implementing all of the Error interface, namely `name` and `message`.

And while it is nice to have an assertion error prototype that can be extended with additional properties, this does not actually work with typescript classes and their contracts unless specifically declared.

A better solution would be to replace the existing typings with either of the below

1. plain object oriented approach

Here, additional properties are defined as properties on the derived classes.

```
declare class AssertionError extends Error {

  public constructor(message: string, ssf?: Function);
}

export = AssertionError;
```

2. or with 'mixin' properties

Here, additional properties can be passed in and the interface allows arbitrary keys for accessing these properties, but there will be no 'type safety'.

```
declare interface AssertionError<T = {}> extends Error {

  [key: string]: any;
}

declare class AssertionError<T = {}> extends Error implements AssertionError<T> {

  public constructor(message: string, props?: T, ssf?: Function);
}

export = AssertionError;
```

One could further reduce the second approach to just

```
declare class AssertionError<T = {}> extends Error {

  public constructor(message: string, props?: T, ssf?: Function);

  [key: string]: any;
}

export = AssertionError;
```

### PR DESCRIPTION
**Pull request by silkentrance**
- refine typescript typings
- add .idea to .gitignore

TODO:

- [x] add additional test for typescript/typings

### CODE REVIEW

**Comment by silkentrance** [comment]
Without the new typings in place, the new "tests" for typescript integration will result in the following errors:

```
> make test


  ==> [Node.js]

  Tests (5)
  1. [pass] construction
  2. [pass] message
  3. [pass] stack
  4. [pass] custom properties
  5. [pass] .toJSON()

  Failures (0)

test/typings.ts:16:45 - error TS2509: Base constructor return type 'AssertionError<T>' is not a class or interface type.

16 class DerivedAssertionError<T = {}> extends AssertionError<T> {
                                               ~~~~~~~~~~~~~~


test/typings.ts:27:13 - error TS2345: Argument of type 'DerivedAssertionError<{ bar: number; }>' is not assignable to parameter of type 'Error'.
  Property 'name' is missing in type 'DerivedAssertionError<{ bar: number; }>'.

27 errors.push(new DerivedAssertionError<{ bar: number }>("msg", { bar: 42 }));
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


make: *** [test-typescript] Error 1
npm ERR! Test failed.  See above for more details.
```

File: .gitignore [hunk 1]
```diff
@@ -17,3 +17,5 @@ node_modules
 npm-debug.log
 
 coverage.html
+
+.idea
```
Comment (keithamus) [inline .gitignore]: This should probably not be part of our gitignore - it's not an artefact this project generates that we don't want to check in, its an editor arteface - and so should be in the user's `~/.gitignore`. 

Can we remove this line please?

Comment (silkentrance) [reply .gitignore]: Of course.

Comment (silkentrance) [reply .gitignore]: I have also remove the extraneous ... impements AssertionError<T> from the class declaration.

**Comment by keithamus** [comment]
Hey @silkentrance thanks for removing the `.idea` line in the `.gitignore` file. Although now it looks like you've commited the `.idea` folder to the project. 

I think you might like to add the `.idea` line to your own gitignore. If you're on a Mac or Linux machine running the command `echo .idea >> ~/.gitignore` will suffice. Then you'll need to modify the commits in this PR to remove the `.idea` folder. I don't know how to add a global gitignore on windows, but [github documentation on gitignore](https://help.github.com/articles/ignoring-files/) or [the official git gitignore documentation](https://git-scm.com/docs/gitignore) may be able to help. Let me know if you run into difficulties with this - we can amend the PR from this side before merging.

**Comment by silkentrance** [comment]
@keithamus argh. darn. fixing this now.

**Comment by silkentrance** [comment]
@keithamus I am unable to add this to a global gitignore as some of my project actually include parts of it in the repository.

**Comment by silkentrance** [comment]
@keithamus ping. what about the status of this?

**Comment by keithamus** [comment]
I'm happy with this, but I want to make sure others are. So someone else on @chaijs/assertion-error will need to merge this.

**Comment by lucasfcosta** [comment]
LGTM. Shall I hit the merge button?

**Comment by keithamus** [comment]
If you're happy with it then feel free to merge 😃

**Pull request merged**